src/core/unit.c

```diff
@@ -1929,6 +1929,102 @@ static bool unit_verify_deps(Unit *u) {
         return true;
 }
 
+static bool unit_can_instantiate(Unit *u, char **ret) {
+        _cleanup_free_ char *t = NULL;
+        unsigned n;
+        int r;
+
+        assert(u);
+        assert(u->manager);
+
+        if (!u->instance)
+                return true;
+
+        if (u->n_max_instances == 0)
+                return true;
+
+        if (!u->manager->instance_counts)
+                return true;
+
+        r = unit_name_template(u->id, &t);
+        if (r < 0)
+                return r;
+
+        n = PTR_TO_UINT(hashmap_get(u->manager->instance_counts, t));
+
+        if (ret)
+                *ret = TAKE_PTR(t);
+        return n < u->n_max_instances;
+}
+
+static int unit_increment_instance_count(Unit *u) {
+        _cleanup_free_ char *t = NULL;
+        unsigned n;
+        int r;
+
+        assert(u);
+        assert(u->manager);
+
+        if (!u->instance)
+                return 0;
+
+        if (u->n_max_instances == 0)
+                return 0;
+
+        r = hashmap_ensure_allocated(&u->manager->instance_counts, &string_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = unit_name_template(u->id, &t);
+        if (r < 0)
+                return r;
+
+        n = PTR_TO_UINT(hashmap_get(u->manager->instance_counts, t));
```

Comment:
i'd sleep more comfortable ad night, if we'd have some overflow test here, i.e. if n is UINT_MAX here return some error, instead of blindly increasing it further